automation/publish_to_maven_local_if_modified.py

```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env python3
+
+# Purpose: Publish android packages to local maven repo, but only if changed since last publish.
+# Dependencies: None
+# Usage: ./automation/publish_to_maven_local_if_modified.py
+
+from pathlib import Path
+import os
+import time
+import hashlib
+import argparse
+import re
+import subprocess
+
+LAST_CONTENTS_HASH_FILE = ".lastAutoPublishContentsHash"
+
+def fatal_err(msg):
+    print(f"\033[31mError: {msg}\033[0m")
+    exit(1)
+
+def run_cmd_checked(*args, **kwargs):
+    """Run a command, throwing an exception if it exits with non-zero status."""
+    kwargs["check"] = True
+    return subprocess.run(*args, **kwargs)
+
+def find_project_root():
+    """Find the absolute path of the project repository root."""
+    cur_dir = Path(__file__).parent
+    while not Path(cur_dir, "LICENSE").exists():
+        cur_dir = cur_dir.parent
+    return cur_dir.absolute()
+
+parser = argparse.ArgumentParser(description="Publish android packages to local maven repo, but only if changed since last publish")
+parser.parse_args()
+
+root_dir = find_project_root()
+if str(root_dir) != os.path.abspath(os.curdir):
+    fatal_err(f"This only works if run from the repo root ({root_dir!r} != {os.path.abspath(os.curdir)!r})")
+
+# Calculate a hash reflecting the current state of the repo.
+
+contents_hash = hashlib.sha256()
+
+contents_hash.update(
+    run_cmd_checked(["git", "rev-parse", "HEAD"], capture_output=True).stdout
+)
+contents_hash.update(b"\x00")
+
+# Git can efficiently tell us about changes to tracked files, including
+# the diff of their contents, if you give it enough "-v"s.
+
+changes = run_cmd_checked(["git", "status", "-v", "-v"], capture_output=True).stdout
+contents_hash.update(changes)
+contents_hash.update(b"\x00")
+
+# But unfortunately it can only tell us the names of untracked
+# files, so we need to slurp their contents in for ourselves.
+
+changes_lines = iter(ln.strip() for ln in changes.split(b"\n"))
+try:
+    ln = next(changes_lines)
+    while not ln.startswith(b"Untracked files:"):
+        ln = next(changes_lines)
+    ln = next(changes_lines) # skip instruction about using `git add`
+    ln = next(changes_lines)
+    while ln:
+        with open(ln, "rb") as f:
+            contents_hash.update(f.read())
+        contents_hash.update(b"\x00")
+        ln = next(changes_lines)
+except StopIteration:
+    pass
+contents_hash.update(b"\x00")
+
+contents_hash = contents_hash.hexdigest()
```

Comment:
Something that occurred to me after doing the a-s version: are the `.gitignore`d files whose contents we should also take into account here? I suspect `local.properties` is one, but there may be others.

Reply:
👍 took your changes from the a-s PR.